hw/util/blinker/include/blinker/blink.h

```diff
@@ -0,0 +1,285 @@
+ *          x / y / 0 / . = Continous blinking on (length) / off (delay)
+ *          x / . / 1 / . = Blink once
+ *          x / y / z / 0 = Repeat z time the on/off sequence and stop
+ *
+ * @details Create the blinking sequence:
+ *          (on[length] / (off[delay] / on[length]){count-1} off[wait])+
+ *
+ * @param length        duration of led on                   (1 .. 255)
+ * @param delay         duration of led off (betwenn led on) (1 .. 255)
+ * @param count         define a group of consecutive on/off (1 .. 15)
+ * @param wait          duration of led off (between group)  (1 .. 255)
+ */
+#define BLINK(length, delay, count, wait)  ((length <<  0) |    \
+                                            (delay  <<  8) |    \
+                                            (wait   << 16) |    \
+                                            (count  << 24))
+
+/**
+ * Always ON
+ */
+#define BLINK_ON                           BLINK(1, 0, 0, 0)
+
+/**
+ * Always OFF
+ */
+#define BLINK_OFF                          BLINK(0, 0, 0, 0)
+
+/**
+ * Blink once and stop
+ *
+ * @param length                duration of 'on' state
+ */
+#define BLINK_ONCE(length)                 BLINK(length, 0, 1, 0)
+
+/**
+ * Continuously blink
+ *
+ * @param length                duration of 'on' state
+ * @param delay                 duration of 'off' state
+ */
+#define BLINK_CONTINUOUS(length, delay)    BLINK(length, delay, 0, 0)
+
+/**
+ * Blink n-time and stop.
+ *
+ * @param length                duration of 'on' state
+ * @param delay                 duration of 'off' state
+ * @param count                 number of blinks
+ */
+#define BLINK_STREAK(length, delay, count) BLINK(length, delay, count, 0)
+
+/* Forward declaration */
+struct blink;
+
+typedef struct blink blink_t;
+
+/*
+ * Compute time for next on/off duration
+ */
+typedef bool (*blink_onoff_t)(blink_t *blink, uint16_t *on, uint16_t *off);
+
+/**
+ * Function to set the led/beeper/... state (on / off)
+ */
+typedef void (*blink_state_t )(bool state);
+
+/**
+ * Blink handler
+ */
+struct blink {
+    /**
+     * Function to drive the led/beeper/... state (on / off)
+     */
+    blink_state_t set_state;
+    /**
+     * Unit of time for blinks (in ticks).
+     */
+    uint32_t time_unit;
+    /**
+     * Blink separator. Delay between two consecutive blink requests.
+     */
+    uint16_t separator;
+    /*
+     * Last blink time
+     */
+    os_time_t last_time;
+    /*
+     * Running 
+     */
+    struct {
+        blink_onoff_t onoff;
+        void *data;
+        uint16_t step;
+        uint8_t flags;
+    } running;
+    /*
+     * Pending 
+     */
+    struct {
+        blink_onoff_t onoff;
+        void *data;
+        uint8_t scheduled;
+    } next;
+    /*
+     * Mutex
+     */
+    struct os_mutex mutex;
+    /*
+     * Callout for 'on'
+     */
+    struct os_callout on_callout;
+    /*
+     * Callout for 'off'
+     */
+    struct os_callout off_callout;
```

Comment:
You could save some space by replacing one of these callouts with a state variable, and keeping state of the 'on/off' state that way.